clippy_lints/src/utils/mod.rs

```diff
@@ -79,30 +79,26 @@ pub fn meets_msrv(msrv: Option<&VersionReq>, lint_msrv: &Version) -> bool {
     msrv.map_or(true, |msrv| !msrv.matches(lint_msrv))
 }
 
-#[macro_export]
 macro_rules! extract_msrv_attr {
     (LateContext) => {
-        fn enter_lint_attrs(&mut self, cx: &rustc_lint::LateContext<'tcx>, attrs: &'tcx [Attribute]) {
-            match get_inner_attr(cx.sess(), attrs, "msrv") {
-                Some(msrv_attr) => {
-                    if let Some(msrv) = msrv_attr.value_str() {
-                        self.msrv = crate::utils::parse_msrv(&msrv.to_string(), Some(cx.sess()), Some(msrv_attr.span));
-                    } else {
-                        cx.sess().span_err(msrv_attr.span, "bad clippy attribute");
-                    }
-                },
-                _ => (),
-            }
-        }
+        extract_msrv_attr!(@LateContext, ());
     };
     (EarlyContext) => {
-        fn enter_lint_attrs(&mut self, cx: &rustc_lint::EarlyContext<'tcx>, attrs: &'tcx [Attribute]) {
-            match get_inner_attr(cx.sess, attrs, "msrv") {
+        extract_msrv_attr!(@EarlyContext);
+    };
+    (@$context:ident$(, $call:tt)?) => {
+        fn enter_lint_attrs(&mut self, cx: &rustc_lint::$context<'tcx>, attrs: &'tcx [rustc_ast::ast::Attribute]) {
+            use $crate::utils::get_unique_inner_attr;
```

Comment:
@Suyash458 This change might be interesting to you, for learning about _macro magic_. I did 3 things here:

1. I removed `macro_export`, because we don't want this macro to be used outside of Clippy
2. I added the `use $crate::utils::get_unique_inner_attr;` statement, so that this function does not have to be imported in very module this is used.
3. I added an "internal rule". Note that there isn't a thing like an "internal rule" officially, but there is a convention that rules starting with `@` should only be used inside the macro definition. With this and the `(, $call:tt)?` thing I was able to deduplicate the code. The `?` operator in a macro means, that this either should appear once or not at all. `tt` means, that `$call` is a token tree, so basically everything that is recognized as a token by the Rust lexer is allowed. In combination this makes passing `()` into the macro possible and can be added anywhere.

Reply:
@flip1995 Thanks, this is interesting! I tried making this generic using trait objects but quickly ran into a lot of issues with lifetimes